app/app-controllers/welcome-screen/welcomeModalController.js

```diff
@@ -90,14 +92,28 @@ function(
         var self = this;
         params = Astro.Utils.extend({forced: false}, params);
 
-        self._secureStore.get('onboarded').then(function(onboarded) {
-            if (onboarded !== 'YES' || params.forced) {
-                self.isShowing = true;
-                self.modalView.show({animated: true});
-                AppEvents.trigger(AppEvents.names.welcomeShown);
-                self._secureStore.set('onboarded', 'YES');
+        return Promise.join(
+            Application.getAppInformation(),
+            self._secureStore.get('installationID'),
+            function(appInfo, previousInstallationID) {
+                // Welcome modal should be triggered when installationID changes
+                // NOTE: appInfo.installationID appears to be different on each app run
+                // on the iOS Simulator, but on real devices they will persist.
+                if (appInfo.installationID !== previousInstallationID || params.forced) {
+                    return new Promise(function(resolve, reject) {
+                        self.isShowing = true;
+                        self.modalView.show({animated: true});
+
+                        // Promise will be resolved when welcome modal is dismissed
+                        AppEvents.on(AppEvents.names.welcomeHidden, function() {
+                            self._secureStore.set('installationID', appInfo.installationID);
+                            resolve();
```

Comment:
Hmm... I feel like I had a conversation with someone before and remembered talking about how the `callback` provided to `Promise.join` would resolve once it hits a return or when it executes to the end of a code path in its scope.

So here the return a `Promise` would be where it resolves whereas if we don't show it... the `callback ` would finish and would `resolve` automatically.

I could definitely make it more explicit in the code though... since it does seem like the promise never resolves.